### PR TITLE
Unpublish comments+articles by user id in API v1

### LINKED ISSUE #17658: Write Endpoint for unpublish all posts and comments
DoD:

AuthZ
AuthN
Audit Trail logging

Documentation

### PR DESCRIPTION
**Pull request by fdocr**
## What type of PR is this? (check all applicable)

- [x] Feature

## Description

New API endpoint that allows admins to unpublish all comments+articles from a user

## Related Tickets & Documents

- Closes #17658

## QA Instructions, Screenshots, Recordings

1. Enable :api_v1 Feature Flag
1. Create an API Key in /settings/extensions
1. Send in a PUT request to /api/users/:id/unpublish with the API Key and application/vnd.forem.api-v1+json in the Accept header

### UI accessibility concerns?

N/A

## Added/updated tests?

- [x] Yes

## [Forem core team only] How will this change be communicated?

- [x] This change does not need to be communicated, and this is why not: WIP project that is still behind FeatureFlag

## [optional] What gif best describes this PR or how it makes you feel?

not looking

### CODE REVIEW

File: app/controllers/api/v1/users_controller.rb [hunk 1]
```diff
@@ -4,6 +4,7 @@ class UsersController < ApiController
       include Api::UsersController
 
       before_action :authenticate!
+      before_action :authorize_super_admin, only: [:unpublish]
```
Comment (jeremyf) [inline app/controllers/api/v1/users_controller.rb:7]: Reading Notes: This is a method in the v0 policy ecosystem.

Editorialization: My preference is that we have policy objects handling this behavior and rely on `authorize(resource)` however that is a larger refactor.

By continuing to use the `authorize_super_admin` we're creating circumstances that mean we have duplication of policy knowledge/enforcement.  This isn't quite a blocker, but instead a "musing outlaid"

Comment (fdocr) [reply app/controllers/api/v1/users_controller.rb:7]: Oh this makes a lot of sense!

> My preference is that we have policy objects handling this behavior and rely on authorize(resource) however that is a larger refactor.

If I understand this correctly, the existing `user_policy` could be used to replace this with an `unpublish` or `admin_unpublish` method that validates the user attempting this action is a super admin.

I can push these changes today 👍🏼

Comment (joshpuetz) [reply app/controllers/api/v1/users_controller.rb:7]: I had the same question/concern in the Suspend User API endpoint PR: https://github.com/forem/forem/pull/18043/files#r918138940

Comment (fdocr) [reply app/controllers/api/v1/users_controller.rb:7]: Whoops, yeah makes sense I overlooked this in both PRs. I'm pushing the proper validation using the policy in both of them soon. Thanks for pointing this out @joshpuetz  and @jeremyf  🙏🏼

File: app/controllers/concerns/api/users_controller.rb [hunk 2]
```diff
@@ -21,5 +21,11 @@ def show
     def me
       render :show
     end
+
+    def unpublish
+      authorize(@user, :unpublish_all_articles?)
```
Comment (fdocr) [inline app/controllers/concerns/api/users_controller.rb:26]: Reusing `:unpublish_all_articles?` that already existed from the Admin dashboard implementation

**Review by joshpuetz — APPROVED** [review]
🥳

**Review by jeremyf — APPROVED** [review]
A provisional approval but our policies are hairy and messy.

**Comment by fdocr** [comment]
> A provisional approval but our policies are hairy and messy.

I'd be happy to sync on this to help move forward the effort to fix this problem. With 30 policy files in the codebase and all the other issues [you've already identified some time ago](https://forem.team/jeremy/forem-article-authorization-discrepencies-45df) I feel like we need a map/check-list of small scoped tasks to solve this.

**Pull request merged**